[TACHYON-613] Login Mechanism: Add GetTachyonLoginUser method using L…

dongche opened the pull request:
JIRA: https://tachyon.atlassian.net/browse/TACHYON-613

After adding OS_LOGIN_MODULE and TACHYON_LOGIN_MODULE, we should provide a static method for invocation to get Tachyon login user. Its implementation will leverage the two added modules. This patch also make some change corresponding with configuration, to make the whole procedure complete.

@ooq , @apc999 , could you please take a look at it? Thanks. Please let me know if there are any confusion.

ooq commented:
Please review this PR. @apc999 @ooq

apc999 commented on common/src/main/java/tachyon/security/UserInformation.java:
```diff
@@ -114,10 +146,89 @@ static String getOsPrincipalClassName() {
+   * Initialize with conf
+   * @param conf the configuration
+   */
+  private static void initialize(TachyonConf conf) {
```
This is confusing. If static methods are used, they meant to do something without maintaining state, which looks to me not the case here --- `sConf` seems a state set externally. If you want to maintain states like a `TachyonConf` or a `AuthTypes`, probably reconsider how to structure this class.

dongche replied:
Done. Re-structure the class and the `TachyonConf` is not set externally.

apc999 commented:
Thanks for the patch!

Here is my high-level feedback: previously I thought this class is simply a collection of constants. It looks to me right now it has states, while it has too many static methods. It makes this class really confusing to me. Could you elaborate what we expect from this class, then we can maybe think about if there is a code structure here.

apc999 commented on common/src/main/java/tachyon/security/UserInformation.java:
```diff
@@ -114,10 +146,89 @@ static String getOsPrincipalClassName() {
+   * Set the static configuration and initialize based on it
+   * @param conf the configuration
+   */
+  public static void setTachyonConf(TachyonConf conf) {
```
Do you have this public static method only for unit-test purpose? If yes, then this is not necessary we have alternative solutions to work around; if no, I am afraind this is probably not how we are suppoed to use a static member as it can be set by an arbitrary external class.

dongche replied:
I guess there might be cases requiring setting the conf dynamically, but still not catch it. Maybe there is not. So this is only for test purpose now.

Is there an example in Tachyon code about the alternative solution to work around? I could take a look and make updates following that way.

ooq commented on common/src/main/java/tachyon/security/UserInformation.java:
```diff
@@ -40,11 +48,35 @@
+  /** The configuration to use */
+  private static TachyonConf sConf;
+  /** The authentication method to use */
+  private static AuthenticationFactory.AuthTypes sAuthType;
```
This is actually related to a closed PR, but can you rename `AuthTypes` to `AuthType`. In java the convention is to use singular for enum, it's quite confusing the other way.

dongche replied:
Done

ooq commented on common/src/main/java/tachyon/security/UserInformation.java:
```diff
@@ -114,10 +146,89 @@ static String getOsPrincipalClassName() {
+   */
+  private static void initialize(TachyonConf conf) {
+    sAuthType = AuthenticationFactory.getAuthTypeFromConf(conf);
+    if (sAuthType == AuthenticationFactory.AuthTypes.NOSASL) {
```
I think we should throw exception for any AuthType except SIMPLE for now.

dongche replied:
Done

ooq commented on common/src/test/java/tachyon/security/LoginModuleTest.java:
```diff
@@ -55,4 +64,47 @@ public void simpleLoginTest() throws Exception {
+  @Test
+  public void initConfTest() throws Exception {
+    // throw exception when initializing conf with "NOSASL"
+    TachyonConf conf = new TachyonConf();
```
Why not simply set it to "NOSASL" for the purpose of this test?

dongche replied:
Done. The class and method names are changed in latest patch. Please refer to `UserInformation.securityEnabledTest()`

ooq commented on common/src/test/java/tachyon/security/LoginModuleTest.java:
```diff
@@ -55,4 +64,47 @@ public void simpleLoginTest() throws Exception {
+      try {
+        UserInformation.setTachyonConf(conf);
+        Assert.fail();
+      } catch (UnsupportedOperationException e) {
```
We are now converting to use the way JUNIT 4 is providing to test exception. Please find examples here:

```
Thanks for adding unit test covering cases with exceptions thrown! this is really helpful for ensuring Tachyon gets solid components.

A style issue here. Tachyon is using JUNIT4 and given this version is used, we use the following convention to unit test thrown exceptions:
http://junit.org/apidocs/org/junit/rules/ExpectedException.html
And here is an example we unit tests with exceptions in Tachyon
https://github.com/amplab/tachyon/blob/master/servers/src/test/java/tachyon/worker/block/BlockMetadataManagerViewTest.java#L110.

```

dongche replied:
Done. Please refer to `UserInformation.securityEnabledTest()`

ooq commented:
@dongche Thanks for the patch. In addition to the comments I left, I also have the same question as @apc999 . I think the confusion here is that we are not entirely clear how this class is going to be used. It would be helpful if you can elaborate more on that, then we can figure out the best way to structure this class.

dongche commented:
Oh, right. I should describe a whole picture for this class at first. Here it is.

- Intention of `UserInformation` (**Bold** part is in the scope of simple auth. Not bold part is out of the scope and is consideration for future wrok.)

  - **The `User` class we already added in TACHYON-600 just represents the user itself. The user falls into two types in concept: one is login user (unique) for Tachyon service, the other is client users (multiple) that connects to Tachyon service by RPC.**

  - Then `UserInformation` instance represents a user in Tachyon from a complete perspective. `UserInformation` wrap `User` and bind other useful state with this `User`, such as `Subject`, `RealUser`, in order to support delegation token(decrease auth frequency for performance ) and impersonation(one user proxy another one).

  - Additionally, it is named as `UserInformation` for clarification in this simple authentication feature scope. Actually, when we move forward to authorization, it will be rename to `UserGroupInformation`, like the Hadoop does. At that time, we may bind more info with the user, such as `Group`, in order to decide permission by checking user and its group.

- `UserInformation` design (only for simple auth stage) - It provide static methods to create kinds of users. Looks like a factory somehow. This class has 4 main part:
  - code block providing JAAS configuration for supporting login. (**Refactor 1: move it to a separate class for clarification.**)
  - static attribute and method about `TachyonConf` and `AuthTypes`, used for switch mode. They should be unique in a Tachyon process.
  - static attribute and method about `loginUser`. This user should be unique in a Tachyon process.
  - static method to create and return a client user.

- How about doing below refactoring?
  - Refactor 1: move the JAAS configuration to a separate class for clarification. (mentioned above)
  - Refactor 2: Simplify the design and impl to only target simple auth scope currently (the bold part above). In the future, we can extend the class when needed. If so, I think `User` is enough for now, and `UserInformation` can be changed to something like `UserFactory`, which provide static methods to create login `User` and remote client `User`.

dongche commented:
@apc999 , @ooq , thanks for your feedback. What's your opinion about the two possible refactor work? And other thought?

Please let me know if any questions. I like this discussion ^_^. It also help me to think more and design the code better.

ooq commented:
@dongche Thanks for the description. To make the discussion easier, let's move to this google doc: https://docs.google.com/document/d/1Ghep_Krsx5W1abXucqns3HYXdxEKtYmqU_caWSQ-F5M/edit?usp=sharing
It's open to edit/comment for everyone.

ooq commented:
@dongche @apc999 Left some comments in the doc. Thanks!

gpang commented on common/src/main/java/tachyon/security/UserInformation.java:
```diff
@@ -40,11 +48,35 @@
+   */
+  UserInformation(Subject subject) {
+    mSubject = subject;
+    mUser = subject.getPrincipals(User.class).iterator().next();
```
Will the iterator always have an item?

dongche replied:
Thanks. Add check on it. Since the structure is changed, the check is added in `UserInformation.login()` method.

ooq commented:
@dongche Left more comments on the doc.

dongche commented:
@ooq , @apc999 , @gpang , Thanks for your review and feedback! I updated the patch to address most of your comment. Could you take a look at it again?

Since the discussion is still going on at the doc, I did not change too much about the structure. It will be in next patch.

dongche commented:
BTW, left some reply on the doc.

ooq commented:
@dongche Thanks for the detailed explanation on the google doc. It helped a great deal in understanding what we want to achieve for user login PRs. I think both refactor will make sense!

ooq commented:
To sum up what we discussed in the doc: 1, I think refactor 1 is really up to you. If we are only targeting at simple mode for now, which means that we only need one JAAS config setting, we can leave JAAS in user information now. That said, it's usually better to have a separate file. 2, We agree that if we don't need to configure things during tachyon runtime, we should try best to config them during construction and make the parameters to be `private final`.

ooq commented:
With respect to the two logging scenarios you talked about, I would view them as, really, two (independent) features: 1, login for the tachyon service so we know who runs the service; 2, login for tachyon client so we know who connects to tachyon. Both are important for security, while 1 matters more for deployment concerns and 2 matters more for multi-tenancy support. I think for this stage, we should only focus on 2 and add 1 after 2 is finished. That is, for now, we should not perform logging on tachyon service, or any logging that is not for the purpose of identifying a user for a tachyon client. Doing one feature after another simply makes the PRs much easier to understand and in turn boosts up code review.

dongche commented:
Thanks for the great summary! @ooq 

Patch updated with changes:
* Split the JAAS configuration and constants to separate files (`TachyonJaasConfiguration` and `TachyonJaasProperties`)
* Simplify `UserInformation` to focus on getting a login user for Tachyon client.
* Refine the javadoc for reviewing easier.
* Split the unit tests to separate files based on the structure change. (`LoginModuleTest` and `UserInformationTest`)

Added some reply to the comments. Fixing for most of them are updated in last commit.

ooq commented on common/src/main/java/tachyon/security/UserInformation.java:
```diff
@@ -15,109 +15,112 @@
+  private static final TachyonConf TACHYON_CONF;
+  /** The authentication method to use. */
+  private static AuthenticationFactory.AuthType sAuthType;
+  /** User instance of the login user in Tachyon process (Tachyon service or Tachyon client) */
```
Let's target only at Tachyon client for now. Given the code should be totally reusable for tachyon service, we can easily add them into master later. In this way, we can also remove some file changes at the master.

dongche replied:
Sure, refine the comments in latest patch. Please refer to `LoginUser` class

ooq commented on common/src/main/java/tachyon/security/UserInformation.java:
```diff
@@ -15,109 +15,112 @@
+  /** The authentication method to use. */
+  private static AuthenticationFactory.AuthType sAuthType;
+  /** User instance of the login user in Tachyon process (Tachyon service or Tachyon client) */
+  private static User sLoginUser;
```
If the user is not changed in the lifespan of this instance. Can we make it private final?

ooq commented on common/src/main/java/tachyon/security/UserInformation.java:
```diff
@@ -15,109 +15,112 @@
   static {
-    OS_LOGIN_MODULE_NAME = getOSLoginModuleName();
-    OS_PRINCIPAL_CLASS_NAME = findOsPrincipalClassName();
+    TACHYON_CONF = new TachyonConf();
```
Can we assign conf during construction, and use private final?

ooq replied:
Can we initialize it directly instead of using this static initializer block?

ooq commented on common/src/main/java/tachyon/security/UserInformation.java:
```diff
@@ -15,109 +15,112 @@
-    OS_LOGIN_MODULE_NAME = getOSLoginModuleName();
-    OS_PRINCIPAL_CLASS_NAME = findOsPrincipalClassName();
+    TACHYON_CONF = new TachyonConf();
+    sAuthType = AuthenticationFactory.getAuthTypeFromConf(TACHYON_CONF);
```
Similarly, can we make AuthType to be private final?

ooq commented on common/src/main/java/tachyon/security/UserInformation.java:
```diff
@@ -15,109 +15,112 @@
-  static String getOsPrincipalClassName() {
-    return OS_PRINCIPAL_CLASS_NAME;
+  @VisibleForTesting
+  static void reset() {
```
Make this method private?

ooq commented on common/src/main/java/tachyon/security/UserInformation.java:
```diff
@@ -15,109 +15,112 @@
-      return SIMPLE;
-    }
+  @VisibleForTesting
+  static void setsAuthType(AuthenticationFactory.AuthType authType) {
```
Make this method private?

apc999 commented:
@dongche thanks for the refactoring (it looks much cleaner!) and providing the explanation with more details. I read the discussion in the gdoc above. Here I have a high level question here regarding `UserInformation`:

Based on its current code, my understanding is that it wraps a  `User` object and a `sAuthType` object and both are static. Correct me if I am wrong:  the main purpose to make them static is to make sure these two values are unique across different references of `UserInformation`, right? 

Here is my thought: static members of a class are supposed to be stateless, which seems not the case in our context---you have setter functions to change them. If `uniqueness` is something we really want to achieve, I think we are using the wrong design pattern and you may want to investigate Singleton for this purpose.

ooq commented:
Agree with @apc999 . From what I see now, the current code structure follows the Singleton design pattern to some extent. My suggestion would be to make methods private whenever possible, and make members to be final and private whenever possible. Thanks!

dongche commented:
@apc999 , @ooq , yes, it looks like singleton. Thanks.

Updated a patch to use Singleton and remove not necessary static members (`sConf` and `sAuthType`). In order to describe the class intention better, this patch rename the class `UserInformation` to `LoginUser` and also the corresponding test class. We use `LoginUser.get()` method to get the unique login user, and it is lazy initialization. The reason of wrapping it with private method `LoginUser.get(TachyonConf)` is for test purpose.

Please let me know your thought on this design and let's make it better.

Below is some reply:
>The main purpose of having them static is to ensure these two values are unique across different references of UserInformation, right?

Yes.

>Whenever you have public setter functions to modify these static members (e.g., setsAuthType and reset), ...

The setter functions like `reset` was intended to be used for unit test and annotated as `VisibleForTesting`. In the new patch, they are removed and use reflection instead. Not sure this workaround is ok? Thoughts?

apc999 commented on common/src/main/java/tachyon/security/LoginUser.java:
```diff
@@ -0,0 +1,122 @@
+   * Check whether Tachyon is running in secure mode, such as SIMPLE, KERBEROS.
+   * @param authType the authentication type in configuration
+   */
+  private static void isSecurityEnabled(AuthenticationFactory.AuthType authType) {
```
NIT: this function seems to do a validation job, rather than returning a boolean value. In this case, `isSecurityEnabled` seems less proper as it indicates returning some boolean by the usual naming convention. How about changing it to `checkSecurityEnabled`?

dongche replied:
Good idea. Done.

ooq commented on common/src/main/java/tachyon/security/LoginUser.java:
```diff
@@ -0,0 +1,122 @@
+   */
+  private static void isSecurityEnabled(AuthenticationFactory.AuthType authType) {
+    //TODO: add Kerberos condition check.
+    if (authType != AuthenticationFactory.AuthType.SIMPLE) {
```
Just a note: per #1176 `AuthType.CUSTOM` is also supported, you might need to change this when merging with #1176 .

dongche replied:
Thanks for reminding. Mark them as TODO in the code.

ooq commented on common/src/main/java/tachyon/security/TachyonJaasConfiguration.java:
```diff
@@ -0,0 +1,60 @@
+  @Override
+  public AppConfigurationEntry[] getAppConfigurationEntry(String appName) {
+    if (appName.equalsIgnoreCase(AuthenticationFactory.AuthType.SIMPLE.getAuthName())) {
+      return SIMPLE;
```
Similarly, you might also want to return simple for `AuthenticationFactory.AuthType.CUSTOM`.

dongche replied:
Yes, marked as TODO

ooq commented on common/src/main/java/tachyon/security/authentication/AuthenticationFactory.java:
```diff
@@ -52,13 +52,22 @@
       return mAuthType;
     }
+
+    public static AuthType getValidAuthType(String authTypeStr) {
```
Javadoc?

dongche replied:
Done

ooq commented on common/src/main/java/tachyon/security/authentication/AuthenticationFactory.java:
```diff
@@ -52,13 +52,22 @@
     }
+
+    public static AuthType getValidAuthType(String authTypeStr) {
+      for (AuthType authType : AuthType.values()) {
```
`AuthType.valueOf(authTypeStr)` should be enough?

dongche replied:
Oh, right, thanks! Done

ooq commented:
@dongche Thanks for the update! I like that the refactoring, the code is much cleaner now. I have left some comments, mostly minor. For the test, I'm wondering if you really need to do reflection to achieve the testing purposes. For example, i think it's fine to make `LoginUser#get(TachyonConf conf)` public as well, then you should be able to change `tachyonConf` in the test without reflection?

apc999 commented:
@dongche , regarding your question about removing `reset`
> The setter functions like reset was intended to be used for unit test and annotated as VisibleForTesting. In the new patch, they are removed and use reflection instead. Not sure this workaround is ok? Thoughts?

Yes, that is a very proper use of reflection in unit tests! thanks for the update!

dongche commented:
@ooq , @apc999 , thanks for your continuous review! Patch updated to address comments.

ooq commented on common/src/main/java/tachyon/security/authentication/AuthenticationFactory.java:
```diff
@@ -52,13 +54,23 @@
+     * @throws java.lang.IllegalArgumentException if the string does not match any type
+     */
+    public static AuthType getValidAuthType(String authTypeStr) {
+      return AuthType.valueOf(authTypeStr.toUpperCase(Locale.ENGLISH));
```
NIT: for debugging purpose, maybe let's catch the exception and throw it again so we can add a more specific error message? (I just don't know what error message would be for this `java.lang.IllegalArgumentException` here. If you think it's good enough, don't bother to fix.)

dongche replied:
The error message is `No enum constant tachyon.security.authentication.AuthenticationFactory.AuthType.WRONG`. I missed using `mThrown.expectMessage()` in the test.

I think it is good to add a more specific error message. Thanks. Currently it is not clear enough. Updated in new patch.

ooq commented:
@dongche The PR looks good. Another question: I see that `NOSASL` is not supported now. As we discussed, for a while, it might be still important to allow some tachyon users to turn off login/authentication, either for performance purpose or if they just want all files to be sharable. How are we going to support that? Are we going to use `NOSASL` as the configuration parameter? Not necessarily related to this PR but just want to throw it out here.

dongche commented:
@ooq Right, we are going to use `NOSASL` of the configuration parameter value to turn off login/auth. Actually, when `NOSASL` is configured, the system behavior will be the same as before. From an implementation perspective, there will be a switch branches when getting a Thrift transport, in which `NOSASL` branch will direct to the original code (`TFramedTransport`). Also login related methods will not be invoked in `NOSASL` related logic. These will be taken care in upcoming Jiras.

BTW, I agree with refining the error message and a patch is updated.

ooq commented:
LGTM, and the plan on enabling `NOSASL` sounds good.
This is cool stuff, thanks!

ooq commented:
A side note: the two opening PRs (#1176 and this one) have conflicts. I would suggest to merge #1176 first.

dongche commented:
Sure, I will rebase this one and change slightly for `CUSTOM`, after #1176 merged.

apc999 commented on common/src/main/java/tachyon/security/LoginUser.java:
```diff
@@ -0,0 +1,122 @@
+   * This constructor should never be invoked.
+   */
+  private LoginUser() {
+
```
NIT: this extra line is not needed.

dongche replied:
Done

apc999 commented on common/src/main/java/tachyon/security/LoginUser.java:
```diff
@@ -0,0 +1,122 @@
+   * @return the login user
+   * @throws java.io.IOException if login fails
+   */
+  public static User get() throws IOException {
```
Not sure if we need this consturctor---we probably will always pass a `TachyonConf` object in the constructor. In general, in current Tachyon codebase, we try to creating creating new `TachyonConf` object in each classes.

dongche replied:
Yes, I was also thinking whether this is needed during dev. Looks a little wired to create a new conf object here.
Removed in the latest patch.

apc999 commented on common/src/main/java/tachyon/security/TachyonJaasConfiguration.java:
```diff
@@ -0,0 +1,61 @@
+  // TODO: add Kerberos_LOGIN module
+  // private static final AppConfigurationEntry KERBEROS_LOGIN = ...
+
+  private static final AppConfigurationEntry[] SIMPLE = new
```
Could you add a comment explaining what the entry `SIMPLE` represents in here? It is not obvious.

dongche replied:
Done

apc999 commented:
Thanks @dongche , this PR turns out a lot cleaner in explaining itself by the code structure. I really appreciate your effort in making this. I left a few very minor comments. Then I will LGTM this PR.

Thanks again.

dongche commented:
@apc999 , updated for the comments. 
I also like current clean code structure, which is realized based on you guys' feedback. Thanks!

apc999 commented:
+1 LGTM

ooq commented:
Jenkins, test this please.

apc999 commented:
@dongche merge failed after  #1176 
@gpang could you take another look on this PR?

gpang commented:
Could you fix the compilation errors? I think there are some issues with PlainSaslServer.java and AuthenticationProviderFactory.java. Thanks!

gpang commented on common/src/main/java/tachyon/security/LoginUser.java:
```diff
@@ -0,0 +1,109 @@
+   * @throws IOException if login fails
+   */
+  private static User login(TachyonConf conf) throws IOException {
+    AuthenticationFactory.AuthType sAuthType = AuthenticationFactory.getAuthTypeFromConf(conf);
```
This is just a local variable, right? If so, then could you rename this to "authType"?

dongche replied:
Yes, thanks. Renamed in the latest patch.

gpang commented:
Jenkins, test this please.

gpang commented:
After the compilation fix and the minor rename, this looks good!

dongche commented:
Patch updated to rebase and fix compilation error.

gpang commented:
LGTM
